src/main/kotlin/org/rust/ide/intentions/SimplifyBooleanExpressionIntention.kt

```diff
@@ -0,0 +1,122 @@
+package org.rust.ide.intentions
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import org.rust.ide.formatter.impl.UNARY_OPS
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.RustTokenElementTypes.*
+import org.rust.lang.core.psi.util.elementType
+import org.rust.lang.core.psi.util.parentOfType
+
+class SimplifyBooleanExpressionIntention : RustElementBaseIntentionAction<SimplifyBooleanExpressionIntention.Context>() {
+    override fun getText() = "Simplify boolean expression"
+    override fun getFamilyName() = text
+
+    data class Context(
+        val expr: RustExprElement
+    )
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        var expr = element.parentOfType<RustExprElement>() ?: return null
+        var lastSimplifiable: RustExprElement? = null
+        if (isSimplifiableExpression(expr))
+            lastSimplifiable = expr
+
+        while (true) {
+            expr = expr.parentOfType<RustExprElement>() ?: break
+            if (isSimplifiableExpression(expr))
+                lastSimplifiable = expr
+        }
```

Comment:
I think this can be rewritten in a more functional style using Kotlin seqnences. Take a look at the `ancestors` method and it's implementation. 

I think it should be possible to do something like `expr.ancestors.takeWhile { it is RustExprElement }.find(isSimplifiableExpression)`.